### PR TITLE
feat: add promise based rendering support

### LINKED ISSUE #9: Add Promise element
At the moment, each element is created synchronously. This prevents us from writing components that fetch data asynchronously, e. g. via a web request. Since markdown is only used for one-off rendering without any interactivity, we don't have to worry about loading states. Therefore we can find a simple solution for asynchronous rendering.

## Describe the solution you'd like

Add a new `md-promise` element which accepts a `Promise` as `children`. Please add a `Promise` element similar to how `Fragment` wraps `md-fragment`.
This should render by resolving the promise. If the promise rejects, reject with the same error.

If you get stuck working on this, I am [available for pair programming](https://calendly.com/bartholomae/open-source-pair-programming).

## Describe alternatives you've considered

React uses hooks and state to manage asynchronous rendering. This seems to be too complicated for this usecase, though.

## Additional context
 
This is based on the breaking change done in #1 so that this change itself no longer is breaking.

When this issue is resolved, we can add additional functionality to [jsx-readme](https://github.com/dbartholomae/jsx-readme) that uses information from the GitHub API to e. g. add an up-to-date list of contributors.

### PR DESCRIPTION
**Pull request by junwen-k**
This is a draft pull request. 

I've been scratching my head for quite some time to get around with this. I've finally manage to complete the async rendering ( Hopefully it is correct ). I am now stuck with how to make the test coverage 100%. Would probably need some advises from you.

P.s: The reason it is called `MdPromise` is to prevent conflict with the javascript keyword `Promise`. Otherwise users will have to always do
```js
import { Promise as MdPromise } from 'jsx-md'; 
```
Hopefully it make sense to you.

Closes #9 

## Checklist:
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: src/createElement.test.ts [hunk 1]
```diff
@@ -7,4 +8,18 @@ describe("createElement", () => {
...
   });
+
+  it("throws an error if a multiple children detected for promise element", () => {
+    expect(() =>
+      // @ts-expect-error - promise element test multiple promises as children
+      createElement(
+        MdPromise,
+        {},
+        new Promise((res) => res("Test")),
+        new Promise((res) => res("Test"))
```
Comment (dbartholomae) [inline src/createElement.test.ts:19]: Please use `Promise.resolve("Test")` instead. Mostly it is shorter, and it took me a while to understand what `res` is :)

File: src/createElement.ts [hunk 2]
```diff
@@ -6,30 +6,53 @@ import {
...
+  }
+
+  if (typeOrComponent === MdPromise) {
+    // Throws error if there is more than one children.
+    if (children.flat().length > 1) {
+      throw new TypeError(
+        `<mdPromise> expected to receive a single promise that resolves to any markdown children.`
+      );
+    }
+    return createPromiseElement(children.flat()?.[0] as MarkdownPromise);
```
Comment (dbartholomae) [inline src/createElement.ts:47]: What if children is null?

Comment (dbartholomae) [reply src/createElement.ts:47]: This is also one of the cases missing for test coverage :)

File: src/createElement.ts [hunk 3]
```diff
@@ -6,30 +6,53 @@ import {
...
   }
 
   if (typeof typeOrComponent === "function") {
-    return createComponentElement(typeOrComponent, attributes, children);
+    return createComponentElement(
+      typeOrComponent as Component,
+      attributes,
+      children as MarkdownNode[]
```
Comment (dbartholomae) [inline src/createElement.ts:54]: These assertions should not be necessary. What error do you get when leaving them out?

File: src/model.ts [hunk 4]
```diff
@@ -29,14 +44,21 @@ export type MarkdownNil = false | null | undefined;
...
 
+/** Promise Markdown type that resolves to nested Markdown type. */
+export type MarkdownPromise = Promise<MarkdownChildren>;
+
 /** Nested Markdown type. */
-export type MarkdownChildren = MarkdownNode | MarkdownChildren[];
+export type MarkdownChildren =
+  | MarkdownNode
+  | MarkdownChildren[]
+  | MarkdownPromise;
```
Comment (dbartholomae) [inline src/model.ts:54]: MarkdownPromise should not be needed here, as it already is part of MarkdownNode via MarkdownElement

File: src/model.ts [hunk 5]
```diff
@@ -29,14 +44,21 @@ export type MarkdownNil = false | null | undefined;
...
 
 /** Helper type for creating Elements that accept other markdown as children. */
 export type PropsWithChildren<
-  AdditionalProps extends unknown = unknown
+  AdditionalProps extends unknown = unknown,
+  Children = MarkdownChildren
 > = AdditionalProps & {
-  children?: MarkdownChildren;
+  children?: Children;
```
Comment (dbartholomae) [inline src/model.ts:61]: This change should not be necessary as MarkdownChildren already contains MarkdownPromise

File: src/render.test.tsx [hunk 6]
```diff
@@ -32,4 +31,42 @@ describe("render", () => {
     );
     expect(await render(<TestComponent attribute="Test" />)).toBe("Test");
   });
+
+  it("throws an error when rendering a promise that rejects with an error", async () => {
+    await expect(
+      render(<MdPromise>{new Promise((res, rej) => rej("invalid"))}</MdPromise>)
```
Comment (dbartholomae) [inline src/render.test.tsx:37]: Please use `Promise.resolve` and `Promise.reject` in this file

File: src/render.tsx [hunk 7]
```diff
@@ -1,18 +1,31 @@
...
+/** @internal */
+function renderNodeAsync(element: MarkdownChildren): Promise<string> {
+  const results = asynchronousRenderFunctions
+    .map((renderFunction) => renderFunction(element, renderNodeAsync))
+    .filter((result) => result !== null);
+  return Promise.resolve(
+    results.length === 0 ? renderNodeSync(element) : (results[0] as string)
+  );
+}
+
```
Comment (dbartholomae) [inline src/render.tsx:27]: I think we can simplify this by renaming "synchronousRenderFunctions" to "renderFunctions" and just adding the new one to the list

File: src/renderFunctions/RenderFunction.ts [hunk 8]
```diff
@@ -3,11 +3,11 @@
...
-import { MarkdownChildren } from "../model";
+import { MarkdownChildren, MarkdownPromise } from "../model";
 
 export interface RenderFunction {
   (
     element: MarkdownChildren,
-    renderer: (el: MarkdownChildren) => string | null
-  ): string | null;
+    renderer: (el: MarkdownChildren) => (string | null) | MarkdownPromise
+  ): (string | null) | MarkdownPromise;
```
Comment (dbartholomae) [inline src/renderFunctions/RenderFunction.ts:12]: I think we can simplify this by always returning promises from all renderers.

File: src/elements/Promise.test.tsx [hunk 9]
```diff
@@ -0,0 +1,26 @@
+/* @jsx MD */
+import MD, { render } from "..";
+
+import { Text, MdPromise } from ".";
+
+describe("MdPromise", () => {
+  it("renders a single promise that resolves to a string", async () => {
+    expect(
+      await render(<MdPromise>{new Promise((res) => res("Test"))}</MdPromise>)
```
Comment (dbartholomae) [inline src/elements/Promise.test.tsx:9]: Please use Promise.resolve and Promise.reject in this file

File: src/elements/Promise.tsx [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+ * Renders a single promise that resolves to a group of tags to markdown asynchronously.
+ * Throws an error if multiple children is provided.
+ * @example
+ *   ```js
+ *   render(<Promise>{new Promise((res) => res("Hello"))}</Promise>)
+ *   ===
+ *   `Hello`
+ *   ```
+ */
+export const MdPromise: Component<PropsWithChildren<
```
Comment (dbartholomae) [inline src/elements/Promise.tsx:15]: Good point that `Promise` is already taken. How about we call it `Await`? Then we could write something like `<Await>{promise}</Await>`

File: src/elements/Promise.tsx [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+  unknown,
+  MarkdownPromise
+>> = ({ children }) => <mdPromise>{children}</mdPromise>;
```
Comment (dbartholomae) [inline src/elements/Promise.tsx:18]: I'm not sure why test coverage for this line is missing. I can have a look at that specifically end of week :)

**Review by dbartholomae — CHANGES_REQUESTED** [review]

File: src/createElement.test.ts (hunk 1 above)
Comment (junwen-k) [reply src/createElement.test.ts:19]: Noted.

File: src/elements/Promise.tsx (hunk 10 above)
Comment (junwen-k) [reply src/elements/Promise.tsx:15]: Noted, will update PR will the change.

File: src/elements/Promise.test.tsx (hunk 9 above)
Comment (junwen-k) [reply src/elements/Promise.test.tsx:9]: Noted.

**Pull request closed without merging**